feat: Editorial review: Attribution reporting API (#33172)
mdn/content: files/en-us/web/api/attribution_reporting_api/index.md

---
title: Attribution Reporting API
slug: Web/API/Attribution_Reporting_API
page-type: web-api-overview
status:
  - experimental
browser-compat: html.elements.a.attributionsrc
---

{{SeeCompatTable}}{{securecontext_header}}{{DefaultAPISidebar("Attribution Reporting API")}}

The **Attribution Reporting API** enables developers to measure conversions — for example when a user clicks an ad embedded on one site and then proceeds to purchase the item over on the vendor's site — and then access reports on those conversions. It does this without relying on third-party tracking cookies.

## Concepts and usage

> Advertisers commonly want to measure conversions so that they can figure out what advertising placements are giving them the greatest return on investment (ROI) and can adjust their strategy to suit. This usually includes capturing data such as:

Reviewer: ```suggestion
Advertisers commonly want to measure conversions so that they can figure out what advertising placements are giving them the greatest return on investment (ROI) and can adjust their strategy accordingly. This usually includes capturing data such as:
```

> Advertisers commonly want to measure conversions so that they can figure out what advertising placements are giving them the greatest return on investment (ROI) and can adjust their strategy accordingly. This usually includes capturing data such as:

Author: I mostly used this, but tweaked it a bit:

> Advertisers commonly want to measure how many users see an ad and then go on to view and purchase a product (conversions). This allows them to figure out what advertising placements are giving them the greatest return on investment (ROI) so they can adjust their advertising strategy accordingly. The process of measuring conversions usually includes capturing data such as:

Traditionally on the web, conversion has been measured using third-party tracking cookies. An ad will typically be embedded on a web page in an {{htmlelement("iframe")}}, which can set a cookie containing information about the user and their interaction with the ad.

Later on, when the user decides to visit the advertiser's site, provided it is from the same domain as the ad, that site can access the third-party cookie set previously by the ad. The advertiser can then associate the data from the ad with their own first-party data to answer questions such as "did the user purchase a product after interacting with an ad for the product from another site?"

This is bad for user [privacy](/en-US/docs/Web/Privacy). At this point, any page from the same domain can get access to that cookie, plus information from sites that embed those pages. A surprisingly large number of parties will be able to access that data, and infer other data about the user based on their browsing habits.

The Attribution Reporting API provides a way to measure ad conversions in a way that protects user privacy.

### How does it work?

Let's illustrate how the Attribution Reporting API works via an example.

> Say we have a online shop, `shop.example` (aka the advertiser), which embeds an ad for one of its products (contained on `ad.shop.example`) on a content site, `news.example` (aka the publisher). The online shop wants to measure how many conversions they get from users interacting with the ad, and then viewing the product page on their site and putting the product into their shopping cart.

Reviewer 2: ```suggestion
Say we have a online shop, `shop.example` (aka the advertiser), which embeds an ad for one of its products (which can be purchased on `ad.shop.example`), on a content site, `news.example` (aka the publisher). The online shop wants to measure how many conversions they get from users interacting with the ad, and then viewing the product page on their site and putting the product into their shopping cart.
```

Author: This isn't quite right. I read this a few times and updated the content to the following, to hopefully avoid the same confusion:

> Say we have a online shop, `shop.example` (aka the advertiser), which embeds an ad for one of its products on a content site, `news.example` (aka the publisher). The ad content is found at `ad.shop.example`.
>
> The online shop owners want to measure how many conversions they get from users interacting with the ad,  viewing the product page on their site, and putting the product into their shopping cart.

Reviewer 2: Whole bunch of review comments I missed. But yes, this is good.

> Say we have a online shop, `shop.example` (aka the advertiser), which embeds an ad for one of its products (contained on `ad.shop.example`) on a content site, `news.example` (aka the publisher). The online shop wants to measure how many conversions they get between users interacting with the ad, and then viewing the product page on their site and putting the product into their shopping cart.

Reviewer 2: "between" doesn't quite work here because it implies you're measuring multiple conversions that happen  between the point that the user interacts with an ad and putting it in their shopping card. You could mean that, but I THINK you're saying that you want to measure how many conversions you get **from** a user interacting with the ad and then subsequently ...

Author: OK, I've changed "between" to "from"

![Image representation of the steps described below](ara-flow.png)

The steps involved are as follows:

1. When a user visits the `news.example` site, the site can register an **Attribution source**. This is an ad-related feature that can be used to detect an interaction with the ad. This feature can be:
   - A link. In this case, an interaction is detected by the user clicking the link (directly via an {{htmlelement("a")}} element, or via a {{domxref("Window.open()")}} call).
   - An image such as a 1x1 transparent pixel. The source is registered when the server responds to the image resource network request.
   - A script. In this case you can indicate interaction with the ad via whatever resource request makes sense for your app (i.e. a {{domxref("fetch")}} or {{domxref("XMLHttpRequest")}}).

> 2. When the action occurs, as determined by the attribution source, associated source data is stored in a private local cache accessible only by the browser. This data includes the contextual and first-party data available to the page and the advertiser, origin of the ad creative, and one or more destinations ([eTLD+1](https://web.dev/same-site-same-origin/#site)s) where you expect the conversion from that ad to occur.

Reviewer 2: What action occurs?
Who is the attribution source? As this point I am a bit confused because it might be a link or it might be the `news.example` site. Can there be multiple sources on the site for each ad, or just one?  

I _think_ that the user cause the site to register an attribution source by clicking on the link, or image, or some element that triggers a script/fetch. But it might be that the browser detects the need to send the headers when the page is loaded (as this is written)?

That then sends a header to the server(s) specfied in href or the attributionSrc to say "hey, I'm a source". The server sends back a JSON to say - yes you are, here is the stuff I want you to record and match if the user goes to this site.

So what you're really saying is that  "the server sends back registration info that ....

Reviewer 2: It might be worth splitting the previous point into two points. The first being the attribution source, which covers how the page sets up the source, The second being about the attribution trigger. Then this action becomes more obvious "when the trigger event is sent" or similar.

Reviewer 2: What is an "ad creative"? I have a fair idea, but this is jargon to tech people.

Reviewer 2: "one or more destinations where you expect the conversion from that ad to occur". YOu might say "such as shop.example.com". I explicitly the site that does the conversion.

Author: OK, so I've made a few changes here:

1. I've changed "action" to "interaction" - what the potential interactions are is explained in point 1. I suspect this typo was what caused a lot of your confusion.
2. I've changed "ad creative" to "the ad tech company that is collecting the conversion data".
3. I've followed your advice in the last point, and updated that bit of the sentence to "where you expect the conversion from that ad to occur, i.e. the advertiser's site(s), for example `shop.example`."

In each case, the feature that has the source (or trigger) registered on it is something that sends a request. If the request contains a valid `Attribution-Reporting-Eligible` header, the response should contain a `Attribution-Reporting-Register-Source` or `Attribution-Reporting-Register-Trigger` header to register the source or trigger. In the case of a link, the request and response will only be sent when the user clicks the link.

The registration request will go to the advertiser, yes. This is also the reporting server. The server that generated the reports has to be same origin with where the registration requests went.

> 3. When the user later visits `shop.example`, this site can register an **Attribution trigger**. Attribution triggers can be registered by loading:

Reviewer 2: So I believe it is `shop.example` that registers an attribution trigger. It isn't clear what causes this to happen, but I think it happens when the page is loaded. So the image loads (say) and that causes a trigger to be registered for the current user.

The trigger then gets set off when the user presses "Buy", say. Right?

Author: OK, I've added a bunch more detail to try to make this clearer, but I've realized that you are bringing up exactly the same questions that I kept asking earlier on when I was trying to document this, and it still isn't clear to me. (I thought I understood it, but now you've asked the questions, I've become confused all over again).

Basically, as I understand it, the fundamental disconnect is:

1. When the user visits the publisher or advertiser site, you register a source or a trigger by sending a request of some kind (e.g. an `<img>` request or a fetch request) with an `Attribution-Reporting-Eligible` header set on it. The advertiser's server then responds with an `Attribution-Reporting-Register-Source` or `Attribution-Reporting-Register-Trigger` as appropriate to register the source or trigger.
2. What *then* needs to happen for the source or trigger to be "set off", and trigger the browser to store the source data (in the case of a registered source), or attempt to match the trigger to a saved source entry (in the case of a registered trigger). Is this the same as point 1 — i.e. does it occur along with the registration...or is it separate, in which case, what is it?

I'm going to go back to the

> 1. When a user visits the `news.example` site, the site can register an **Attribution source**. This is an ad-related feature that represents an interaction with the ad. In each case, the feature will send a request along with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution source, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Source")}} header. The feature can be, for example:
>    - A link. In this case, the interaction is the user clicking on the link (directly via an {{htmlelement("a")}} element, or via a {{domxref("Window.open()")}} call). The source is registered via the response to the navigation request.
>    - An image such as a 1x1 transparent pixel. In this case, the interaction is the user visiting the page. The source is registered when the image loads, i.e. when the server responds to the image request.
>    - A fetch request (i.e. a {{domxref("fetch()")}} or {{domxref("XMLHttpRequest")}}). In this case the interaction can be specified as whatever makes sense for your app — for example the fetch request could be invoked by a `click` or `submit` event. The source is registered once the response comes back.

Reviewer 2: This is so much better. My only issue with it is "When a user visits the `news.example` site, the site can register an **Attribution source**.". That implies that visiting the site is enough to register the source, and also that only one source can be registered. My thinking is that even if you have an image and that is loaded, it probably represents just one ad. You probably have multiples of images if there are multiple ads right? The use of "feature" is also a bit clunky.

It would be good to make it clear that the attribution source might apply to multiple ads. Something like:

```suggestion
1. When a user visits the `news.example` site, the site can register an **Attribution source** for each interaction that the user has with an ad. There are several ways that a user can interact with ads on the page. Each interaction results in a request being sent with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution source, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Source")}} header. The attribution source that the user interacts with can be one of:
   - A link. In this case, the interaction is the user clicking on the link (directly via an {{htmlelement("a")}} element, or via a {{domxref("Window.open()")}} call). The source is registered via the response to the navigation request.
   - An image such as a 1x1 transparent pixel. In this case, the interaction is the user visiting the page. The source is registered when the image loads, i.e. when the server responds to the image request.
   - A fetch request (i.e. a {{domxref("fetch()")}} or {{domxref("XMLHttpRequest")}}). In this case the interaction can be specified as whatever makes sense for your app — for example the fetch request could be invoked by a `click` or `submit` event. The source is registered once the response comes back.
```

   - An image, for example a 1x1 conversion pixel.
   - Any network resource (via {{domxref("fetch")}} or {{domxref("XMLHttpRequest")}}).

> 4. When the attribution trigger is set off (for example, the user clicks the "Add to cart" button on `shop.example`), the browser attempts to match the attribution trigger to a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:

Reviewer 2: Perhaps overcomplicated at this level, but it might be worth saying that the attribution trigger is the HTTP header Xxxx. I.e. a little on the mechanics. Even if as an aside.

Author: In my rewrite; I've covered a bit about the mechanics in 3.

> 2. When the source interaction occurs, the associated source data (sent in the {{httpheader("Attribution-Reporting-Register-Source")}} header) is stored in a private local cache accessible only by the browser. This data includes the contextual and first-party data available to the page and the advertiser, the origin of the ad tech company that is collecting the conversion data, and one or more destinations ([eTLD+1](https://web.dev/same-site-same-origin/#site)s) where you expect the conversion from that ad to occur (i.e. the advertiser's site(s), for example `shop.example`).

Reviewer 2: Now you are referring to Attribution-Reporting-Register-Source you don't need awkward terms like source interaction.

```suggestion
2. When attribution source registration is confirmed, the source data returned in the {{httpheader("Attribution-Reporting-Register-Source")}} header is stored in a private local cache accessible only by the browser. This data includes the contextual and first-party data available to the page and the advertiser, the origin of the ad tech company that is collecting the conversion data, and one or more destinations ([eTLD+1](https://web.dev/same-site-same-origin/#site)s) where you expect the conversion from that ad to occur (i.e. the advertiser's site(s), for example `shop.example`).
```

Author: Right. I've deliberately been keeping away from using "registration" as much as possible, because of the confusion we discussed earlier in this process — the registration is the same thing as the source interaction, but a registration sounds like a setup step, which will then require a later interaction, whereas an interaction sounds like the thing we want.

I've updated the first bit to "attribution source interaction occurs" in case that helps. I've made your later change (source data returned), as that is much better than what I had.

> 3. When the user later visits `shop.example`, this site can register an **Attribution trigger**. This is an feature that represents an interaction with the shop page indicating that a conversion has occurred (for example, the user clicks the "Add to cart" button on `shop.example`). In each case, the feature will send a request along with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution trigger, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Trigger")}} header. The feature can be, for example:

Reviewer 2: Ditto for when trigger registration occurs, though not as important. But maybe something like::

```suggestion
3. When the user later visits `shop.example`, this site can register an **Attribution trigger** when an interaction occurs that indicates that a conversion has occurred (for example, the user clicks the "Add to cart" button on `shop.example`). When such an interaction occurs, the browser will send a request along with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution trigger, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Trigger")}} header. The feature can be, for example:
```

Author: I very nearly used this all, but modified the very last sentence:

> 3. When the user later visits `shop.example`, this site can register an **Attribution trigger** when an interaction occurs that indicates that a conversion has occurred (for example, the user clicks the "Add to cart" button on `shop.example`). When such an interaction occurs, the browser will send a request along with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution trigger, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Trigger")}} header. The attribution trigger can be, for example:

> 4. When trigger interaction occurs, the browser attempts to match the attribution trigger to a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:

Reviewer 2: ```suggestion
4. When the trigger attribution completes, the browser attempts to match the data from the [Attribution-Reporting-Register-Trigger](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger) to a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:
```

Author: Better. I modified it a little bit:

> When the trigger attribution completes, the browser attempts to match the data from the [Attribution-Reporting-Register-Trigger](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger) header with a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:

> 4. When trigger interaction occurs, the browser attempts to match the attribution trigger to a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:

Reviewer 2: > For a successful match

Are these the only criteria for a match (I don't think so). We should link to other information about matching. This is my confusion around all the noise stuff

Author: The matching criteria in the "Registering attribution triggers" guide basically says the same thing. AFAIUI, noise had nothing to do with the browser's ability to make a match — it just fuzzes the data points so that individuals can't be identified from their data as easily.

  - : Causes completion of the registration of an attribution source _and_ triggers the browser to store the associated source data (as provided in the {{httpheader("Attribution-Reporting-Register-Source")}} response header) when the `open()` method completes. Note that `Window.open()` calls cannot be used to register attribution triggers.

## HTML elements

- {{htmlelement("a")}}, {{htmlelement("img")}}, and {{htmlelement("script")}} — the `attributionsrc` attribute
  - : Specifies that you want the browser to send an {{httpheader("Attribution-Reporting-Eligible")}} header along with the associated resource request. On the server-side this header is used to trigger sending an {{httpheader("Attribution-Reporting-Register-Source")}} or {{httpheader("Attribution-Reporting-Register-Trigger")}} header in the response. When registering an attribution source, this is required; when registering an attribution trigger it is only required if you want to specify a separate registration server to the resource the `src` attribute is pointing at. Note that `<a>` elements cannot be used to register attribution triggers.

## HTTP headers

- {{httpheader("Attribution-Reporting-Eligible")}}

>   - : Used to indicate that the response corresponding to the request is eligible to take part in attribution reporting, by registering either an attribution source or trigger.

Reviewer 2: This needs work, both here and in the item. A responds to a request is an HTTP header. It isn't eligible to do anything.

I think that this header "indicates an intent to a [reporting server?] to register the site as either an attribution source or trigger"

Author: "Eligible" is the exact word the spec/API uses. I don't think it is completely crazy — it means if the response to this request includes an appropriate `Attribution-Reporting-Register-*` header, then the effect will be to register the feature that sent the request as a source/trigger.

If we randomly sent a `Attribution-Reporting-Register-*` header in response to a request that didn't include a `Attribution-Reporting-Eligible` header, the registration wouldn't work. It wouldn't be eligible.

Reviewer 2: If specs were intelligible, you and I would be unnecessary :-). I think this is better, given the context. 
```suggestion
  - : HTTP request that indicates that the corresponding response is eligible to register an attribution source or trigger.
```

Author: Yup, yours is a lot better ;-)

In my next commit I've updated this on the API landing page, and I've also updated the corresponding text on the HTTP header page.

- {{httpheader("Attribution-Reporting-Register-Source")}}

>   - : Included as part of a response to a request that included an `Attribution-Reporting-Eligible` header, this is used to register an attribution source.

Reviewer 2: I would consider inverting because it is better to say what it does without much context then provide the context. 
Here and above I use ` [page | site | whatever]` as a choose your own adventure - I think a page is what is registered, but it might be something else. We should be very clear on this somewhere (and we are not, unless I am missing some high level info)

```suggestion
  - : HTTP response that registers a [page | site | whatever] as an attribution source. This is included as part of a response to a request that included an `Attribution-Reporting-Eligible` header.
```

Author: Done, and done. I have used the term "page feature", to keep consistent with the wording I've used in the rewrite of the main API landing page. So, an ad isn't a source, it is a feature of that ad, be it a link, an image, or whatever.

This is bad for user [privacy](/en-US/docs/Web/Privacy). At this point, any page from the same domain can get access to that cookie, plus information from sites that embed those pages. A surprisingly large number of parties will be able to access that data, and infer other data about the user based on their browsing habits.

The Attribution Reporting API provides a way to measure ad conversions in a way that protects user privacy.

### How does it work?

Let's illustrate how the Attribution Reporting API works via an example.

Say we have a online shop, `shop.example` (aka the advertiser), which embeds an ad for one of its products on a content site, `news.example` (aka the publisher). The ad content is found at `ad.shop.example`.

The online shop owners want to measure how many conversions they get from users interacting with the ad, viewing the product page on their site, and putting the product into their shopping cart.

![Image representation of the steps described below](ara-flow.png)

The steps involved are as follows:

> 1. When a user visits the `news.example` site, the site can register an **Attribution source** for each interaction that the user has with an ad. There are several ways that a user can interact with ads on the page. Each interaction results in a request being sent with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution source, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Source")}} header. The attribution source can be, for example:

Reviewer 2: I am not 100% certain that "the site can register" is correct/helpful because it implies that `news.example` is what registers the attribution source. 

My understanding is that news.example embeds an ad in an iframe from shop.exmaple, which might be an image element or whatever. So it is the ad that causes the attribution source to be registered.

This matters because I had a disconnect as to how the same-origin for the reporting and destinations could work if the news.example was a different origin.

Reviewer 2: Though in note further down:

> > **Note:** These requirements provide privacy protection, but also flexibility — the source _and_ trigger can potentially be situated on the top-level site, or embedded in an {{htmlelement("iframe")}}.

So the requirement is that the source/destination/reporting be same origin. But I stand by my comment above that in this example "the site" is confusing. I'd also update the note to "the source and trigger must be on the same origin" - often this will be an iframe from the ad site, but it could be a top level site ... blah

Author: I get what you are saying here. I've updated the paragraph to the following, to try to untangle this:

> When a user visits the `news.example` site, an **attribution source** can be registered for specific user interactions with the embedded ad. There are several ways that a user can interact with ads on the page. For an ad interaction to register an attribution source, it must send a request with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution source. Registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Source")}} header. The attribution source can be, for example:

WDYT?

Reviewer 2: I think this bit is good:

> When a user visits the news.example site, an attribution source can be registered for specific user interactions with the embedded ad. 

This bit, is a little confusing, because the "it" in "it must send a request " is ambiguous.

> For an ad interaction to register an attribution source, it must send a request with an

Author: Right, I see what you mean. I've updated it to:

> "For an ad interaction to register an attribution source, the ad must send a request with an..."

To make it more explicit.

   - An image such as a shopping cart icon or a 1x1 transparent tracking pixel. In this case, the interaction is the user visiting the page. The trigger is registered when the image loads, i.e. when the server responds to the image request.
   - A fetch request (i.e. a {{domxref("fetch()")}} or {{domxref("XMLHttpRequest")}}). In this case the interaction can be specified as whatever makes sense for your app — for example the fetch request could be invoked by a `click` or `submit` event. The trigger is registered once the response comes back.
4. When the trigger attribution completes, the browser attempts to match the data from the [Attribution-Reporting-Register-Trigger](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger) header with a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:
   - On a `destination` specified in the source's associated data.
   - Same-origin with the request that specified the source registration.

>      > **Note:** These requirements provide privacy protection, but also flexibility — the source _and_ trigger can potentially be situated on the top-level site, or embedded in an {{htmlelement("iframe")}}.

Reviewer 2: See comment above about this note - I think the iframe is the most likely use case, so should be presented first

Author: I've moved this note to the "Registering attribution triggers" article, but your point still stands. I've swapped the order over, as suggested. The note now reads:

> **Note:** These requirements provide privacy protection, but also flexibility — the source _and_ trigger can potentially be embedded in an {{htmlelement("iframe")}} or situated in the top-level site.

> 5. If a match is made, the browser sends report data to an endpoint on a reporting server typically owned by the ad tech provider where it can be securely analyzed. The data is not accessible by the site the ad is placed on, or the advertiser site, or any other site except for the site hosting the reporting endpoint. These reports can be either:

Reviewer 2: > typically owned by the ad tech provider where it can be securely analyzed. The data is not accessible by the site the ad is placed on, or the advertiser site, or any other site except for the site hosting the reporting endpoint.

The point I think you're making here is that, unlike with cookies, the data is only available to the specific site you send it to - no data leakage. Might be worth making that clear since my first though was "so what - the advertiser owns all those sites so they can see the info anyway".

Author: Gotcha. Your sentence gets to the point much better, so I've used it with a couple of modifications. I've updated the sentence starting with "The data..." to:

> Unlike with cookies, the data is only available to the specific site you send it to - there will be no data shared elsewhere.

   - An image such as an advertising banner or a 1x1 transparent tracking pixel. In this case, the interaction is the user visiting the page. The source is registered when the image loads, i.e. when the server responds to the image request.
   - A fetch request (i.e. a {{domxref("fetch()")}} or {{domxref("XMLHttpRequest")}}). In this case the interaction can be specified as whatever makes sense for your app — for example the fetch request could be invoked by a `click` or `submit` event. The source is registered once the response comes back.
2. When the attribution source interaction occurs, the source data returned in the {{httpheader("Attribution-Reporting-Register-Source")}} header is stored in a private local cache accessible only by the browser. This data includes the contextual and first-party data available to the page and the advertiser, the origin of the ad tech company that is collecting the conversion data, and one or more destinations ([eTLD+1](https://web.dev/same-site-same-origin/#site)s) where you expect the conversion from that ad to occur (i.e. the advertiser's site(s), for example `shop.example`).

> 3. When the user later visits `shop.example`, this site can register an **Attribution trigger** when an interaction occurs that indicates that a conversion has occurred (for example, the user clicks the "Add to cart" button on `shop.example`). When such an interaction occurs, the browser will send a request along with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution trigger, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Trigger")}} header. The attribution trigger can be, for example:

Reviewer 2: Too many occurs
```suggestion
3. When the user later visits `shop.example`, this site can register an **Attribution trigger** when an interaction indicates that a conversion has occurred (for example, the user clicks the "Add to cart" button on `shop.example`). The browser will then send a request along with an {{httpheader("Attribution-Reporting-Eligible")}} header to indicate that the response is eligible to register an attribution trigger, and registration will be completed if the response includes an appropriate {{httpheader("Attribution-Reporting-Register-Trigger")}} header. The attribution trigger can be, for example:
```

   - An image such as a shopping cart icon or a 1x1 transparent tracking pixel. In this case, the interaction is the user visiting the page. The trigger is registered when the image loads, i.e. when the server responds to the image request.
   - A fetch request (i.e. a {{domxref("fetch()")}} or {{domxref("XMLHttpRequest")}}). In this case the interaction can be specified as whatever makes sense for your app — for example the fetch request could be invoked by a `click` or `submit` event. The trigger is registered once the response comes back.
4. When the trigger attribution completes, the browser attempts to match the data from the [Attribution-Reporting-Register-Trigger](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger) header with a source data entry saved in the private local cache (see 2.). For a successful match, the trigger must be:
   - On a `destination` specified in the source's associated data.

>    - Same-origin with the request that specified the source registration.

Reviewer 2: Below (search on "When the user interacts with the attribution trigger, the browser attempts to match the trigger") you also mention data trigger as a matching condition. You might have chosen not to mention it here deliberately

Author: Ah, so you mean the requirements stated in the "Registering attribution triggers" article are different? To fix this, I have removed the requirements information from this article completely, made sure all the information is available in the other article (I moved the "These requirements provide privacy protection, but also flexibility..." note), and linked to it from here. 

Does that work for you?

Reviewer 2: Correct. On re-reading I see my explanation was poor. Your fix is good.
